### PR TITLE
JDK-8293877: Rewrite MineField test

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review an update to convert the remaining langtools shell tests, in `test/langtools/tools/javac/Paths/*` from shell to Java code.

This will be an awkward/difficult/tedious (choose your adjective) review.  The old files have been deleted, and the new code is different enough to defeat any `diff` program.  That being said, I have done what I can to provide a line-for-line translation as best I can, in the parts of the code that matter.  It may help to use tools to display the old code and new code side-by-size, in an ad-hoc fashion, and manually scrolling the versions together.

The old shell tests were written in a stylized and clever manner to reduce the verbiage of writing in shell script. The new tests are written in a differently stylized manner, using API methods to provide much of the same level of convenience.

Of the six tests, 5 have an obvious name translation (the hyphen is dropped from `Class-Path.sh` and `Class-Path2.sh`, and the cryptically named `wcMineField.sh` is renamed to the less cryptic `WildcardMineField.java`.

The supporting API in the new `Util.java` class is somewhat similar to the older `Util.sh` code, but is much less of a line-for-line translation.

Where possible, tools are invoked using the tool's `ToolProvider` API, as compared to exec-ing a separate process. A separate process is still always necessary for the main Java launcher, and is sometimes needed for Java ... when javac should be executed in a different directory, or with different env variables, or when using the "class path wildcard" feature.

On the use of `cleanup` methods... all the tests here call a `cleanup()` method as part of initial setup, and after the test is complete. Such cleanup is unnecessary beforehand when using jtreg (because jtreg guarantees an empty scratch directory when a test is run) and can be inconvenient afterwards, if you want to debug why a test-case has failed, although for sure it is arguably good to do if the test has passed.  But any change would be a difference in the old and new code, and should arguably be done separately.  So, the existence and use of `cleanup` methods is as before.

Some code was commented out in the original tests, because various options and system properties became unavailable in JDK 9. The commented out code is preserved, if only to provide visual anchors between the old and new versions. It may be desirable to eventually clean out the commented-out code.

The old code supported use of `${TESTTOOLVMOPTS}` and `${TESTVMOPTS}` although it is not clear how effective such support was, or how much it was used. The new code does _not_ support these environment variables, or their equivalent system properties. Given that the new code uses the `ToolProvider` API where possible, it is not clear how (or how easily) we could integrate support for those options, and whether it is worth the effort.  

Needless to say, all the new tests pass. For positive test cases, that is definitely reassuring. For negative test cases, it is somewhat harder to verify that each new test case fails for the same reason that the old test case failed, not least because of the lack of info provided by the old tests.  For this reason, it is assumed it is sufficient to have confidence in the translation of all the test cases.

Separately, there is a minor fix in ToolBox, to address an NPE. The fix is small enough and obvious enough to include here, as compared to having a separate JBS issue and PR.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293877](https://bugs.openjdk.org/browse/JDK-8293877): Rewrite MineField test

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)
 * [Martin Buchholz](https://openjdk.org/census#martin) (@Martin-Buchholz - **Reviewer**)
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10366/head:pull/10366` \
`$ git checkout pull/10366`

Update a local copy of the PR: \
`$ git checkout pull/10366` \
`$ git pull https://git.openjdk.org/jdk pull/10366/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10366`

View PR using the GUI difftool: \
`$ git pr show -t 10366`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10366.diff">https://git.openjdk.org/jdk/pull/10366.diff</a>

</details>

### CODE REVIEW

**Comment by Martin-Buchholz** [comment]
Explicit OS command invocation and ToolProvider invocation both seem important to test.
IIRC there were times when they were not in fact equivalent in effect.
Consider library infrastructure to run both variants and ensure results are identical.

I recall being slightly proud when writing the first iteration of this test - it allowed a very terse test description in shell.
But terseness is not universally admired, and even I (in my dotage) have joined the camp of writing tests in more verbose junit style.

**Comment by jonathan-gibbons** [comment]
> Explicit OS command invocation and ToolProvider invocation both seem important to test. IIRC there were times when they were not in fact equivalent in effect. Consider library infrastructure to run both variants and ensure results are identical.
> 

@Martin-Buchholz 
Thanks for the feedback. While I understand what you are suggesting, I think it is potentially for a different test.

The goal here was to try and do a reasonably direct translation of the original code from shell to Java. And OK, I did do some minor optimizations to use javac's `ToolProvider` where possible, to save exec-ing processes unnecessarily.

Rewriting the test to do both (where that is even possible) would either be a significant rewrite or maybe involve running all of the test twice, in both modes. 

And, for javac, it might be even worse, because there is the `javax.tools.JavaCompiler` API as well.

There is no reason to suspect the functionality might be semantically different in these different ways to invoke javac. Once you get down into the compiler, the code paths are the same.   The difference is the _availability_ of features ... such as the _classpath wildcard_ feature, which is only available when using the native launcher.   So, if we were to write anything more in the way of tests, it would be to test the specified difference in availability of the features, without necessarily exercising the full functionality and comparing everything against the launcher for the Java runtime, `java`.

One other outcome from the overall work here: as a result of (re)discovering the _classpath wildcard_ feature, we realized that there were enough differences between the native launcher and the `ToolProvider` interface (and `JavaCompiler`) to justify a spec update for the `idk.compiler` module. This is being tracked in [JDK-8294546](https://bugs.openjdk.org/browse/JDK-8294546).

File: test/langtools/tools/javac/Paths/ClassPath.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        moveFiles(List.of("A.jar", "B.zip"), "jars/.");
+
+        expectPass(JAVAC, "-cp jars/A.jar Main.java");
+        expectPass(JAVA, "-cp jars/A.jar${PS}. Main");
+
+        expectPass(JAVAC, "-cp jars/B.zip Main.java");
+        expectPass(JAVA, "-cp jars/B.zip${PS}. Main");
+
+        /*
+         * The following lines are commented out in the original,
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/Paths/ClassPath.java:130]: I think these old comments could be removed

**Review by vicente-romero-oracle — APPROVED** [review]
I have gone through all the changes, I think semantically the original tests have been preserved as closed as possible. After a first review iteration I think it could make sense to clean up a bit an remove comments etc

**Comment by jonathan-gibbons** [comment]
> I have gone through all the changes, I think semantically the original tests have been preserved as closed as possible. After a first review iteration I think it could make sense to clean up a bit an remove dead code etc

Yeah, I'm inclined to agree that the time has come to remove the dead code.

**Review by vicente-romero-oracle — APPROVED** [review]
dead code removal looks good

**Comment by Martin-Buchholz** [comment]
It's clear that javac team intends to have ToolProvider and command line invocation be very equivalent (good!) but I have memories of it being quite tricky to convert from one to the other, and I would be wary of lingering bugs, especially in the path handling code.

I care more than most about the performance of test runs, but for command line flags, I would always at least test them via subprocess invocation.

**Review by Martin-Buchholz — APPROVED** [review]
I spent a few nostalgic minutes looking over the translation into java.  Looks good.

I regret not having added more comments about why particular invocations were expected to pass or fail.  

I was introduced to junit after writing these.  I might have tried to do more junitization if I was doing the translation.

Despite having implemented wildcards, I was never keen on the feature, and haven't used it myself in the Real World.

File: test/langtools/tools/javac/Paths/ClassPath.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+ * @build toolbox.ToolBox Util ClassPath
+ * @run main ClassPath
+*/
+
+
+/*
+ * Converted from Class-Path.sh, originally written by Martin Buchholz.
+ *
+ * For the last version of the original, Class-Path.sh, see
+ * https://github.com/openjdk/jdk/blob/jdk-19%2B36/test/langtools/tools/javac/Paths/Class-Path.sh
```
Comment (jddarcy) [inline test/langtools/tools/javac/Paths/ClassPath.java:39]: Please change this and other github URLs to use "https://git.openjdk.org/jdk" as the prefix, e.g.:

https://git.openjdk.org/jdk/blob/jdk-19%2B36/test/langtools/tools/javac/Paths/Class-Path.sh

**Review by jddarcy — APPROVED** [review]

**Review by lahodaj — APPROVED** [review]
Looks good to me.

**Comment by jonathan-gibbons** [comment]
> It's clear that javac team intends to have ToolProvider and command line invocation be very equivalent (good!) but I have memories of it being quite tricky to convert from one to the other, and I would be wary of lingering bugs, especially in the path handling code.
> 
> I care more than most about the performance of test runs, but for command line flags, I would always at least test them via subprocess invocation.

The comment is noted.   Generally, the javac code is_much_ better that it has been in times past, with better use of better abstractions. That being said, I agree there are different code paths in the outer layers of `javac` and it is important to be aware of the differences when introducing new options.

File: test/langtools/tools/javac/Paths/ClassPath.java (hunk 2 above)
Comment (jonathan-gibbons) [reply test/langtools/tools/javac/Paths/ClassPath.java:39]: Darn!  missed this; will have to fix separately

**Pull request closed without merging**